src/Common/src/TypeSystem/IL/ILProvider.cs

```diff
@@ -15,11 +15,14 @@ namespace Internal.IL
 {
     internal sealed class ILProvider : LockFreeReaderHashtable<MethodDesc, ILProvider.MethodILData>
     {
-        private PInvokeILProvider _pinvokeILProvider;
+        private readonly PInvokeILProvider _pinvokeILProvider;
 
-        public ILProvider(PInvokeILProvider pinvokeILProvider)
+        private readonly bool _readyToRunMode;
```

Comment:
This doesn't feel like a good direction. If you look around this class, you'll see it's doing a bunch of CoreRT specific intrinsic expansion - not just the PInvokes. I really don't want us to end up with a dozen `if (_readyToRunMode)` and then not being able to quickly tell what methods are applicable when (especially if we also start to write ReadyToRun-specific intrinsics too).

I think we'll want to make ILProvider configurable.

* Right now, ILProvider is conflating two things: caching, and providing IL. We'll want to split these two off. So I would start with making ILProvider not derive from LockFreeReaderHashTable and becoming abstract, with the `GetMethodIL` method becoming virtual.
* There would be a CoreRTILProvider that derives from it. It would pretty much keep all the existing logic from the existing ILProvider, unchanged (except for the caching part). I would kick out the `_pinvokeILProvider` logic from it, completely (mostly because we'll not have a PInvokeILProvider available when we'll need to instantiate the CoreRTILProvider - see later).
* There will also be a ReadyToRunILProvider that pretty much only handles the `MethodForInstantiatedType`, `InstantiatedMethod` and default `EcmaMethod` logic. Later, we might need to expand the `Unsafe` intrinsics, but most of the stuff in the existing CoreRTILProvider is not applicable.
* The specific `ILProvider` will be instantiated in the driver (ilc.exe's Program.cs) and a new method will be added to `CompilationBuilder` (e.g. `UseILProvider(ILProvider)`) to set it up before compilation. Compilation builder will pass it as a new parameter down to the specific `Compilation`.
* On the `Compilation` class, make the `PInvokeILProvider` property virtual so that you can make it return `null` in R2R mode.
* Make a new `ILCache` class by stealing the caching logic from the old `ILProvider` (you can make it a nested class under `Compilation`). `ILCache` will take an `ILProvider` as a constructor parameter and cache its results.
* `Compilation` will take the `ILProvider` it got in the constructor, put it under the cache and use it in the `GetMethodIL` method.
* If the cache returns null in `GetMethodIL`, it will check if the method is a PInvoke and if so, it will query the PInvokeILProvider (unless it's null). We can potentially put the PInvoke IL provider under a different cache, but I'm not sure it's necessary. It might be pretty easy if we make the PInvokeILProvider class derive from the PInvokeProvider class though.

I can do this refactoring in the mainline branch and we can integrate it to R2R if you would prefer me doing this. It touches a lot of files, but it's not complicated. The "touching a lot of files" part might be a good enough reason to do this in `master` either way though.

Reply:
Your proposal sounds great to me. As you have a very clear idea what should be done and much better overall knowledge of the codebase, it would be definitely great if you could do the refactoring - I suspect that's going to be much easier than coaching me to try to fit your ideas. Thank you!

Reply:
Sure - #6366 is how much I could do in master. Once that merges, I'll do a merge to r2r branch and implement the R2R part. That should fix some of the failures you're seeing.

Reply:
Awesome, thank you! To my limited understanding your change looks just great, I completely agree this is much cleaner than my initial attempt.